lib/tree.rb

```diff
@@ -16,40 +16,99 @@ def initialize
     @root = nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(log n)
+  # Space Complexity: O(log n)
+  def add_helper(curr, key, value)
+    return TreeNode.new(key, value) if curr.nil?
+
+    if key < curr.key
+      curr.left = add_helper(curr.left, key, value)
+    else
+      curr.right = add_helper(curr.right, key, value)
+    end
+
+    return curr
+  end
+
   def add(key, value)
-    raise NotImplementedError
+    @root = TreeNode.new(key, value) if @root.nil?
+
+    return add_helper(@root, key, value)
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(log n)
+  # Space Complexity: O(log n)
   def find(key)
-    raise NotImplementedError
+    curr = @root
+
+    until curr.nil?
+      if key > curr.key
+        curr = curr.right
+      elsif key < curr.key
+        curr = curr.left
+      else
+        return curr.value
+      end
+    end
+
+    return nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
+  def inorder_helper(curr, list)
+    return list if curr.nil?
+
+    inorder_helper(curr.left, list)
+    list << {key: curr.key, value: curr.value}
+    inorder_helper(curr.right, list)
```

Comment:
Just for clarity on what's being returned, this is a minor suggestion.

```suggestion
    inorder_helper(curr.right, list)
    return list
```